packages/core/src/component.test.ts

```diff
@@ -0,0 +1,311 @@
+            }
+            const component = new TestComponent()
+            component.connectedCallback()
+
+            expect(component.getState()).toStrictEqual(state)
+            expect(component.selector).toHaveBeenCalledWith(state)
+            expect(subscribeWithSelector).toHaveBeenCalledTimes(1)
+        })
+        it("calls bindActionCreators when actionCreators are provided", () => {
+            const actionCreators = {
+                a: () => ({type: "a"})
+            }
+            const store = createEmptyStore(null)
+            store.getState = jest.fn()
+            class TestComponent extends Component {
+                store = store
+                actionCreators = actionCreators
+                bindActionCreators = jest.fn()
+            }
+            const component = new TestComponent()
+            component.connectedCallback()
+
+            expect(subscribeWithSelector).toHaveBeenCalledTimes(0)
+            expect(store.getState).toHaveBeenCalledTimes(0)
+            expect(component.bindActionCreators).toHaveBeenCalledTimes(1)
+        })
+        it("calls callRender and connected in connectedCallback", () => {
+            class TestComponent extends Component {
+                callRender = jest.fn()
+                connected = jest.fn()
+            }
+            const component = new TestComponent()
+            component.connectedCallback()
+
+            expect(component.callRender).toHaveBeenCalledTimes(1)
+            expect(component.connected).toHaveBeenCalledTimes(1)
+        })
+        it("does not call unsubscribe if not set", () => {
+            class TestComponent extends Component {
+                disconnected = jest.fn()
+            }
+            const component = new TestComponent()
+            component.disconnectedCallback()
+
+            expect(component.disconnected).toHaveBeenCalledTimes(1)
+        })
+        it("does calls unsubscribe if is set", () => {
+            const store = createEmptyStore(null)
+            class TestComponent extends Component {
+                disconnected = jest.fn()
+                unsubscribe = jest.fn(store.subscribe(null))
+            }
+            const component = new TestComponent()
+            component.disconnectedCallback()
+
+            expect(component.disconnected).toHaveBeenCalledTimes(1)
+            expect(component.unsubscribe).toHaveBeenCalledTimes(1)
+        })
+        it("does not call attributeChangedCallback when values are undefined or equal", () => {
+            class TestComponent extends Component {
+                propChanged = jest.fn()
+
+                attributeChangedCallback(name: string, oldValue: string, newValue: string) {
+                    super.attributeChangedCallback(name, oldValue, newValue)
+                }
+            }
+            const component = new TestComponent()
+            const value = "value1"
+
+            component.attributeChangedCallback("name1", undefined, undefined)
```

Comment:
There are several consts that I see: "name1", "undefined"

Reply:
I don't understand - what should i change?

Reply:
Can you see the repeated const string values? I suppose you can make consts instead of strings

Reply:
fixed